File: demo/src/main/java/kaufland/com/demo/Application.java
package kaufland.com.demo;

import android.support.annotation.Nullable;
import android.util.Log;

import com.couchbase.lite.Blob;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseConfiguration;

import java.util.ArrayList;
import java.util.Arrays;

import kaufland.com.coachbasebinderapi.PersistenceConfig;
import kaufland.com.coachbasebinderapi.PersistenceException;
import kaufland.com.couchbaseentityconnector.Couchbase2Connector;
import kaufland.com.demo.entity.ProductEntity;
import kaufland.com.demo.entity.UserCommentWrapper;

public class Application extends android.app.Application {


    private static final String TAG = Application.class.getName();
    public static final String DB = "mydb_db";

    private Database mDatabase;

    @Override
    public void onCreate() {
        super.onCreate();

        deleteDbIfExists();

        PersistenceConfig.configure(new Couchbase2Connector() {
            @Override
            protected Database getDatabase(String name) {
                if (DB.equals(name)) {
                    return Application.this.getDatabase();
                }
                throw new RuntimeException("wrong db name defined!!");
            }
        });
        createMockArticle();
    }

    private void deleteDbIfExists() {

        try {
            getDatabase().delete();
            mDatabase = null;
        } catch (CouchbaseLiteException e) {
            Log.e(TAG, "failed to clear Database", e);
        }
    }

    @Nullable
    public Database getDatabase() {
        try {
            if (mDatabase == null) {
                DatabaseConfiguration config = new DatabaseConfiguration(getApplicationContext());
                mDatabase = new Database(DB, config);
            }
            return mDatabase;
        } catch (CouchbaseLiteException e) {
            Log.e(TAG, "failed to get Database", e);
            return null;
        }
    }

    private void createMockArticle() {
        try {
            ProductEntity.create().setName("Beer").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("very awesome"), UserCommentWrapper.create().setComment("tasty")))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
            ProductEntity.create().setName("Beer (no alcohol)").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("very bad"), UserCommentWrapper.create().setComment("not tasty").setAge(99)))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
            ProductEntity.create().setName("Wodka").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("feeling like touch the sky")))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
            ProductEntity.create().setName("Gin").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("hipster drink but great"), UserCommentWrapper.create().setComment("tasty!!!")))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
            ProductEntity.create().setName("Apple").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("mhmhmh tasty!"), UserCommentWrapper.create().setComment("dont like it")))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
            ProductEntity.create().setName("Tomatoes").
                    setComments(new ArrayList<>(Arrays.asList(UserCommentWrapper.create().setComment("don't like there color"), UserCommentWrapper.create().setComment("worst experience ever!!")))).
                    setImage(new Blob("image/jpeg", getResources().openRawResource(R.raw.ic_kaufland_placeholder))).
                    save();
        } catch (PersistenceException e) {
            e.printStackTrace();
        }
    }


}

File: demo/src/main/java/kaufland/com/demo/Application.kt
package kaufland.com.demo

import android.util.Log
import com.couchbase.lite.Blob
import com.couchbase.lite.CouchbaseLiteException
import com.couchbase.lite.Database
import com.couchbase.lite.DatabaseConfiguration
import kaufland.com.coachbasebinderapi.PersistenceConfig
import kaufland.com.coachbasebinderapi.PersistenceException
import kaufland.com.coachbasebinderapi.TypeConversion
import kaufland.com.coachbasebinderapi.TypeConversionErrorWrapper
import kaufland.com.couchbaseentityconnector.Couchbase2Connector
import kaufland.com.demo.customtypes.GenerateClassName
import kaufland.com.demo.customtypes.GenerateClassNameConversion
import kaufland.com.demo.entity.ProductEntity
import kaufland.com.demo.entity.UserCommentWrapper
import kotlin.reflect.KClass

class Application : android.app.Application() {

    private var mDatabase: Database? = null

    val database: Database?
        get() {
            return try {
                if (mDatabase == null) {
                    val config = DatabaseConfiguration(applicationContext)
                    mDatabase = Database(DB, config)
                }
                mDatabase
            } catch (e: CouchbaseLiteException) {
                Log.e(TAG, "failed to get Database", e)
                null
            }
        }

    override fun onCreate() {
        super.onCreate()

        deleteDbIfExists()
        PersistenceConfig.configure(object : Couchbase2Connector() {
            override fun getDatabase(name: String): Database {
                if (DB == name) {
                    return database!!
                }
                throw RuntimeException("wrong db name defined!!")
            }

            override val typeConversions: Map<KClass<*>, TypeConversion>
                get() {
                    val mutableMapOf =
                        mutableMapOf<KClass<*>, TypeConversion>(GenerateClassName::class to GenerateClassNameConversion())
                    mutableMapOf.putAll(super.typeConversions)
                    return mutableMapOf
                }

            override fun invokeOnError(errorWrapper: TypeConversionErrorWrapper) {
                if (errorWrapper.exception is java.lang.ClassCastException) {
                    Log.e(
                        TAG,
                        "Data type manipulated: Tried to cast ${errorWrapper.value} into ${errorWrapper.`class`}"
                    )
                } else throw errorWrapper.exception
            }
        })
        createMockArticle()
    }

    private fun deleteDbIfExists() {

        try {
            database!!.delete()
            mDatabase = null
        } catch (e: CouchbaseLiteException) {
            Log.e(TAG, "failed to clear Database", e)
        }
    }

    private fun createMockArticle() {
        try {
            ProductEntity.create().builder().setName("Beer").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("very awesome").exit(),
                    UserCommentWrapper.create().builder().setComment("tasty").exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
            ProductEntity.create().builder().setName("Beer (no alcohol)").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("very bad").exit(),
                    UserCommentWrapper.create().builder().setComment("not tasty").setAge(99).exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
            ProductEntity.create().builder().setName("Wodka").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("feeling like touch the sky")
                        .exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
            ProductEntity.create().builder().setName("Gin").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("hipster drink but great")
                        .exit(),
                    UserCommentWrapper.create().builder().setComment("tasty!!!").exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
            ProductEntity.create().builder().setName("Apple").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("mhmhmh tasty!").exit(),
                    UserCommentWrapper.create().builder().setComment("dont like it").exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
            ProductEntity.create().builder().setName("Tomatoes").setComments(
                listOf(
                    UserCommentWrapper.create().builder().setComment("don't like there color")
                        .exit(),
                    UserCommentWrapper.create().builder().setComment("worst experience ever!!")
                        .exit()
                )
            ).setImage(Blob("image/jpeg", resources.openRawResource(R.raw.ic_kaufland_placeholder)))
                .exit().save()
        } catch (e: PersistenceException) {
            e.printStackTrace()
        }
    }

    companion object {
        private val TAG = Application::class.java.name
        const val DB = "mydb_db"
    }
}
